src/lib/group.go

```diff
@@ -343,6 +343,76 @@ func group_add(ctx pf.PfCtx, args []string) (err error) {
 	return
 }
 
+//Function to allow TG admins and sysadmins to request a reset of the passowrd
+//for a member of a group that they posses an admin bit for.
+//This function provides the admin making the request the nominator portion
+//of the reset token , and emails the user poriton to the user, thus resolving the
+//bad nominator problem.
+
+func group_pw_reset(ctx pf.PfCtx, args []string)(err error) {
+	var nom_portion string
+	var user_portion string
+	var user_email pf.PfUserEmail
+	var reset_user_email pf.PfUserEmail
+	var pw pf.PfPass
+
+	//Capture Reset user e-mail
+	reset_user_email,err = ctx.SelectedUser().GetPriEmail(ctx, false)
+
+	//Establish that user current user is a group admin for selected group.
+	grpname := args[0]
+	username := args[1]
+	err = ctx.SelectGroup(grpname,pf.PERM_GROUP_ADMIN)
+	if err != nil {
+		return
+	}
+
+	//Select the target user.
+	err = ctx.SelectUser(username, pf.PERM_USER_VIEW)
+	if err != nil {
+		//Return same error to prevent information leakage.
+		err = errors.New("Unable to select user")
+		return
+	}
+
+	grp := ctx.SelectedGroup()
+	user := ctx.SelectedUser()
+
+	//Validate that the user is a member of the target group.
+	var ismember bool
+	ismember, _, _, err = grp.IsMember(user.GetUserName())
+	if !ismember {
+		//Return same error to prevent information leakage.
+		err = errors.New("Unable to select user")
+		return
+	}
+
+	user_email, err = user.GetPriEmail(ctx, true)
+	if err != nil {
+		return
+	}
+
+	user_portion, err = pw.GenPass(16)
+	if err != nil {
+		return
+	}
+
+	nom_portion, err = pw.GenPass(16)
+	if err != nil {
+		return
+	}
+
+	err = Mail_PassResetUser(ctx, user_email, true, reset_user_email, user_portion)
+	if err != nil {
+		return
+	}
+
+	ctx.OutLn("Nominator portion of the password: "+nom_portion)
+	err = user.SetRecoverToken(ctx, user_portion+nom_portion)
```

Comment:
I hope that error isn't important...
shouldn't setting the token happen before mailing, and check that the set worked (no error) before mailing out and telling the user: "nom portion is"

Reply:
Prior to this change we sent mail first in case that failed. I agree that it is a better failure mode to set the code and fail the e-mail.  (fixed)